File: src/main/java/me/udnek/fnafu/map/location/LocationSingle.java
package me.udnek.fnafu.map.location;

import com.google.common.base.Preconditions;
import org.bukkit.Location;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LocationSingle implements LocationData {

    private Location location;
    private boolean frozen = false;

    public LocationSingle add(Location location){
        Preconditions.checkArgument(!frozen, "Location is frozen");
        this.location = location;
        return this;
    }
    public LocationSingle add(double x, double y, double z){
        add(new Location(null, x, y, z, 0, 0));
        return this;
    }
    public LocationSingle add(double x, double y, double z, float yaw, float pitch){
        add(new Location(null, x, y, z, yaw, pitch));
        return this;
    }

    public List<Location> getAll(){
        return new ArrayList<>(Collections.singleton(location));
    }

    public void setOrigin(Location origin){
        Preconditions.checkArgument(!frozen, "Location is frozen");
        LocationData.locationAddOrigin(location, origin);
        frozen = true;
    }


    public Location getFirst(){
        return location.clone();
    }

    public Location getRandom(){
        return getFirst();
    }

    @Override
    public Location get(int n) {return getFirst();}

    public int getSize() { return (location == null ? 0 : 1); }


    public LocationSingle center(){
        Preconditions.checkArgument(!frozen, "Location is frozen");
        Location center = location.toCenterLocation();
        location.set(center.getX(), center.getY(), center.getZ());
        return this;
    }
    public LocationSingle centerFloor(){
        Preconditions.checkArgument(!frozen, "Location is frozen");
        Location center = location.toCenterLocation();
        location.set(center.getX(), center.getBlockY(), center.getZ());
        return this;
    }


    public static LocationSingle from(double x, double y, double z){
        return new LocationSingle().add(x, y, z);
    }

    public static LocationSingle from(double x, double y, double z, float yaw, float pitch){
        return new LocationSingle().add(x, y, z, yaw, pitch);
    }
    public static LocationSingle from(Location location){
        return new LocationSingle().add(location);
    }

}

File: src/main/kotlin/me/udnek/fnafu/map/location/LocationSingle.kt
package me.udnek.fnafu.map.location

import com.google.common.base.Preconditions
import me.udnek.fnafu.misc.setOrigin
import me.udnek.fnafu.misc.toCenterFloor
import org.bukkit.Location

class LocationSingle : LocationData {
    private var frozen = false

    private var location: Location

    override val all: List<Location>
        get() = ArrayList(setOf(location))

    override val first: Location
        get() = location.clone()

    override val random: Location
        get() = first

    override val size: Int = 1

    constructor(location: Location){
        this.location = location.clone()
    }

    constructor(x: Double, y: Double, z: Double, yaw: Float, pitch: Float) : this(Location(null, x, y, z, yaw, pitch))

    constructor(x: Int, y: Int, z: Int, yaw: Float, pitch: Float): this(x.toDouble(), y.toDouble(), z.toDouble(), yaw, pitch)

    constructor(x: Double, y: Double, z: Double): this(x, y, z, 0f, 0f)

    constructor(x: Int, y: Int, z: Int): this(x.toDouble(), y.toDouble(), z.toDouble())

    override fun get(n: Int): Location {
        return first
    }

    override fun setOrigin(origin: Location) {
        Preconditions.checkArgument(!frozen, "Location is frozen")
        location.setOrigin(origin)
        frozen = true
    }

    override val center: LocationSingle
        get() {
            Preconditions.checkArgument(!frozen, "Location is frozen")
            val center = location.toCenterLocation()
            location.set(center.x, center.y, center.z)
            return this
        }

    override val centerFloor: LocationSingle
        get() {
            Preconditions.checkArgument(!frozen, "Location is frozen")
            location = location.toCenterFloor()
            return this
        }

    override val head: LocationSingle
        get() {
            Preconditions.checkArgument(!frozen, "LocationList is frozen")
            location.add(0.0, 1.7, 0.0)
            return this
        }
}
